File: week07/7/[241014](Leet119)_Pascal Triangle.cpp
/*문제
Given an integer rowIndex, return the rowIndexth (0-indexed) row of the Pascal's triangle.

In Pascal's triangle, each number is the sum of the two numbers
*/

/*문제풀이 1: recursive
    1. base case : rowIndex == 0 || rowIndex == 1 
    2. recursive case : row[i] = prev[i-1] + prev[i]
*/

/*문제풀이2 : Iterative
    1. prev, temp를 활용한다. prev를 이용한 temp 한 사이클이 끝나면, prev = temp로 update.
    2. 초기 prev = []
    3. temp를 i번 update.
        for loop : i = 0 ~ rowIndex
            temp[0] = [1]
            temp[i] = 1
            중간값 업데이트
            for loop : j = 1 ~ (i-1)
                temp[j] = prev[j-1] + prev[j]
            prev = temp
    4. retrn temp
*/

#include <iostream>
#include <vector>
using namespace std;

//문제풀이1 : recursive
class Solution{
public:
    vector<int> getRow(int rowIndex){
        vector<int> result(rowIndex+1,1); // 1로 rowIndex+1개만큼 요소 로 초기화

        //base case
        if (rowIndex == 0 || rowIndex == 1) {
            return result;
        }
        
        //recursive
        vector<int> prev = getRow(rowIndex-1);
        for (int i = 1; i < rowIndex; i++){
            result[i] = prev[i-1] + prev[i];
        }

        return result;
    }
};

//문제풀이2 : iterative
class Solution{
public:
    vector<int> getRow(int rowIndex){
        vector<int> prev = {};

        for (int i = 0; i <= rowIndex; i++){
            vector<int> temp(i+1, 0);
            temp[0] = 1;
            temp[i] = 1;
            for (int j = 1; j < i; j++){
                temp[j] = prev[j-1] + prev[j];
            }
            prev = temp;
        }
        return prev;
    }
};

//문제풀이3 : iterative
class Solution{
public:
    int* getRow(int rowIndex){
        int* cur = new int[rowIndex + 1];  // 새로운 정수 배열을 동적으로 할당
        cur[0] = 1;  // 첫 번째 요소는 항상 1

        for (int i = 1; i <= rowIndex; i++) {
            cur[i] = 1;  // 매 행의 마지막 요소는 항상 1
            for (int j = i - 1; j > 0; j--) {
                cur[j] = cur[j - 1] + cur[j];  // 역순으로 중간 값을 업데이트
            }
        }
        return cur;  // 동적으로 할당된 배열의 포인터 반환
    }
};


int main(){
    int rowIndex = 3;
    Solution sol1;
    vector<int> result = sol1.getRow(rowIndex);

    cout << "[";
    for (int i : result) {
        cout << i << " " ;
    }
    cout << "]";

    return 0;
}

File: week07/7/[241014](Leet119)_Pascal Triangle.py
"""문제
row Index를 받았을 때, 해당 row의 pascal triangle값을 list로 return.
"""

"""풀이1 : 재귀적 방식
    1.(Base) rowIndex가 A일때, list[0], list[A]는 1이고.
    2. (Recursive)나머지는 Row[n][i] = Row[n-1][i-1] + Row[n-1][i]
"""
"""풀이2 : Iterative 방식
    1. Row[n][i] = Row[n-1][i-1] + Row[n-1][i] 식을 사용하는데... Row[n]를 temp로, Row[n-1]을 prev로.
    2. prev = [] 에서 시작해서, prev를 사용해서 temp를 만들고, temp를 prev로 바꿔주는 과정을 반복하다가, 원하는 시점의 temp가 나왔을 때 return
"""

"""풀이3 : Dynamic Programming
    1. 삼각형형태로 DP table을 만든다고 생각하면 됨.
    2. (rowIndex + 1)개의 row를 파스칼 삼각형. 이 삼각형을 DPtable이라 보면 되겠음. t = [1] * (rowIndex+1)
    3. 하나의 row를 채워가는 for loop
        for i in range(rowIndex+1) :
            t[i] = [1]*(i+1)
            for j in range(1, i):
                t[i][j] = t[i-1][j-1] + t[i-1][j]
    return t[rowIndex]
"""

# 0 [1]
# 1 [1, 1]
# 2 [1,2,1] i1 = j0+j1
# 3 [1,3,3,1] i1 = j0+j1, i2 = j1+j2
# 4 [1,4,6,4,1] i3 = j2+j3
# 5 [1,5,10,10,5,1]

# 풀이1 : 재귀
def getRow(rowIndex)->list:
    #Base case
    if rowIndex == 0 :
        row = [1]
        return row
    elif rowIndex == 1:
        row = [1,1]
        return row
    
    #Recursive
    row = [1] * (rowIndex+1) # rowIndex+1 길이만큼.
    row_pre = getRow(rowIndex-1)
    for i in range(1, rowIndex):
        row[i] = row_pre[i-1] + row_pre[i]
    return row
    

#풀이2 : Iterative
def getRow(rowIndex) :
    prev = []
    for i in range(rowIndex+1):
        temp = [1]* (i+1)
        for j in range(1,i):
            temp[j] = prev[j] + prev[j-1]
        prev = temp
    
    return temp

#풀이3 : DP
def getRow(rowIndex):
    dpT = [1] * (rowIndex+1)
    for i in range(rowIndex+1):
        dpT[i]  = [1]*(i+1) #i번째 row를 채우는 과정
        for j in range(1, i):
            dpT[i][j] = dpT[i-1][j-1] + dpT[i-1][j]
    return dpT[rowIndex]



#test
for i in range(6):
    print(f"{i} Index의 getRow 결과는 {getRow(i)}")
